byway/src/main/webapp/routepage.js

```diff
@@ -49,27 +63,145 @@ function calcRoute(directionsService, directionsRenderer, start, end) {
   });
 }
 
-function addToStops(stop){
-  console.log("add" + stop.placename + "to the ArrayList in the servlet");
+/** Clear the stops panel in the html */
+function clearStops(){
+  const stopList = document.getElementById('stop-list');
+  if(stopList != null){
+    stopList.innerText = ""; // clear list
+  }
 }
 
-function getRecs() {
-  fetch('/api/recs')
+/** Render stop list  */
+function renderStops(stop, fromDatastore){
+  const stopList = document.getElementById('stop-list');
+  // add to stop array in the js 
+  if(fromDatastore){
+    stops.push(stop);
+  }
+  let btn = document.createElement('button');
+  btn.id = stop.id;
+  btn.innerText = stop.placename;
+  btn.className =  "btn rec-btn";
+  btn.addEventListener("click", function() {
+    deleteFromStops(stop); 
+  });
+  stopList.appendChild(btn);
+}
+
+/** Get the new list of stops from datastore onload */
+function getStopsOnload(){
+  clearStops();
+  fetch('/api/stop')
   .then(response => response.json())
-  .then((recs) => {
-    const recList = document.getElementById('rec-list');
-    recList.innerText = ""; // clear old comments 
-    recs.forEach((stop) => {
-      let btn = document.createElement("button");
-      btn.innerText = stop.placename;
-      btn.setAttribute("class", "btn rec-btn");
-      btn.addEventListener("click", function() {
-        addToStops(stop);
-      });
-      recList.appendChild(btn);
+  .then((stopsResponse) => {
+    stopsResponse.forEach((stop)=>{
+      renderStops(stop, true);
     });
+  })
+}
+
+/** Get the new list of stops locally */
+function getStopsList(){
+  clearStops();
+  // re-render list synchronously
+  stops.forEach((stop)=>{
+    renderStops(stop, false);
+  })
+}
+
+/** Add stop locally and to datastore */
+function addToStops(stop){
```

Comment:
Instead of having an add and remove function, it would be better to have a single `updateStops()` method. Here, you would simply pass the entire list  of stops rather than the single item you want to add or remove. This objects are simplistic enough that you can can pass the whole list to the server

Reply:
I used a simplified Trip class to simulate storing and getting destinations in this updateStops() method. I also made it so you can either add or remove a place all in the same function, but I could split it up for readability. Please let me know what is best.